lib/types-simple.nix

```diff
@@ -0,0 +1,457 @@
+    # A: `null` has strong connotations as the “always existing”
+    # alternative value; of course in a unityped language like
+    # nix this is moot, but here we take the chance to throw out
+    # this harmful idea (the “million dollar mistake”).
+    check = v: v == {};
+  };
+
+  # the type with two inhabitants
+  bool = mkScalar {
+    name = "bool";
+    description = "boolean";
+    check = builtins.isBool;
+  };
+
+  # a nix string
+  string = mkScalar {
+    name = "string";
+    description = "string";
+    check = builtins.isString;
+  };
+
+  # a signed nix integer
+  int = mkScalar {
+    name = "int";
+    description = "integer";
+    check = builtins.isInt;
+  };
+
+  # a nix floating point number
+  float = mkScalar {
+    name = "float";
+    description = "float";
+    check = builtins.isFloat;
+  };
+
+  # helper for descriptions of recursive types
+  # TODO: descriptions need to assume t is a type,
+  # which is only true for Fix Type. How to make nice?
+  describe = t: t.description or "<non-type>";
+
+  # list with children of type t
+  # list bool:
+  #   [ true false false ]
+  # list (attrs unit):
+  #   [ { a = {}; } { b = {}; } ]
+  #   []
+  list = t: mkRecursive {
+    name = "list";
+    description = "list of ${describe t}";
+    check = builtins.isList;
+    # each child gets named by its index, starting from 0
+    each = l: builtins.listToAttrs
+      (lib.imap0 (i: v: lib.nameValuePair (toString i) v) l);
+    nested = t;
+  };
+
+  # attrset with children of type t
+  # attrs int: { foo = 23; bar = 42; }
+  # attrs (attrs string):
+  #  { foo.bar = "hello"; baz.quux = "x"; }
+  #  { x = { y = "wow"; }; }
+  attrs = t: mkRecursive {
+    name = "attrs";
+    description = "attrset of ${describe t}";
+    check = builtins.isAttrs;
+    each = lib.id;
+    nested = t;
+  };
+
+  # TODO: nonempty list and attrs
+
+  # product type with fields of the specified types
+  # product { x = int; y = unit; }:
+  #   { x = 23; y = {}; }
+  #   { x = 42; y = {}; }
+  # product {}: <- yeah, that’s isomorphic to unit
+  #   { }
+  # product { foo = void; }:
+  #   just kidding. :)
+  product = fields: productOpt { req = fields; opt = {}; };
+
+  # product type with the possibility of optional fields
+  # actually the more generic type of product, BUT:
+  # code with a fixed number of fields is less brittle.
+  # choose wisely.
+  # productOpt { req = {}; opt = { a = unit; b = int; }:
+  #   { }
+  #   { a = {}; }
+  #   { a = {}; b = 23; }
+  productOpt = { req, opt }: mkBaseType {
+    name = "product";
+    description = "{ " +
+      lib.concatStringsSep ", "
+        (  lib.mapAttrsToList (n: t: "${n}: ${describe t}") req
+        ++ lib.mapAttrsToList (n: t: "[${n}: ${describe t}]") opt)
+      + " }";
+    check = v:
+      let reqfs = builtins.attrNames req;
+          optfs = builtins.attrNames opt;
+          vfs   = builtins.attrNames v;
+      in builtins.isAttrs v &&
+      (if opt == {}
```

Comment:
I'd use `intersectAttrs` (especially because it's a primop) here instead of converting to list and use `subtractLists`.

Reply:
To check for `then reqfs == vfs` in the line below, I’d have to use `builtins.intersectAttrs v req == req`, which has to do a deep equality check for `req` (or `v` if I turn the arguments around).